src/dcqc/filesystems/synapsefs.py

```diff
@@ -0,0 +1,640 @@
+
+from dcqc.filesystems.remote_file import RemoteFile
+
+if TYPE_CHECKING:
+    from fs.info import RawInfo
+
+
+@contextmanager
+def synapse_errors(path):
+    try:
+        yield
+    except SynapseFileNotFoundError:
+        raise ResourceNotFound(path)
+    except SynapseHTTPError as err:
+        message = err.args[0]
+        if "does not exist" in message:
+            raise ResourceNotFound(path)
+        elif "already exists" in message:
+            raise DirectoryExists(message)
+        elif "File size must be at least one byte" in message:
+            message = (
+                "File was empty during an attempted upload to Synapse. "
+                "Synapse doesn't support empty files, so upload was skipped."
+            )
+            warn(message)
+        else:
+            raise  # Raise original exception as is
+
+
+class SynapseFS(FS):
+    """Construct a Synapse filesystem for
+    `PyFilesystem <https://pyfilesystem.org>`_
+
+    Args:
+        root (str): Synapse ID for a project or folder.
+        auth_token (str, optional): Synapse personal access token.
+            Defaults to None.
+        synapse_args (dict, optional): Dictionary of
+            arguments to pass to the ``Synapse()`` class.
+            Defaults to None.
+    """
+
+    DELIMITER = "/"
+    CWD_SYMBOL = "."
+    PARENT_DIR = ".."
+
+    NULL_BYTE = b"\x00"
+
+    SYNID_REGEX = re.compile(r"syn[0-9]+")
+
+    SUPPORTED_TYPES = {
+        "file": File,
+        "folder": Folder,
+        "project": Project,
+    }
+
+    DEFAULT_SYNAPSE_ARGS = {
+        "silent": True,
+    }
+
+    _meta = {
+        "case_insensitive": False,
+        "invalid_path_chars": "\0",
+        "network": True,
+        "read_only": False,
+        "thread_safe": True,
+        "unicode_paths": False,
+        "virtual": False,
+    }
+
+    # TODO: Add option to initialize without root, which allows you
+    #       to open any Synapse ID that you have access to. The only
+    #       requirement is that every path starts with a Synapse ID.
+    def __init__(
+        self,
+        root: Optional[str] = None,
+        auth_token: Optional[str] = None,
+        synapse_args: Optional[dict] = None,
+    ) -> None:
+        super(SynapseFS, self).__init__()
+        self.auth_token = auth_token
+        self.synapse_args = synapse_args or self.DEFAULT_SYNAPSE_ARGS
+        self._local = threading.local()
+        self.root = self._process_root(root)
+
+    @property
+    def synapse(self) -> Synapse:
+        """Construct a thread-local Synapse client.
+
+        Returns:
+            Synapse: Authenticated Synapse client
+        """
+        if not hasattr(self._local, "synapse"):
+            synapse = Synapse(**self.synapse_args)
+            synapse.login(authToken=self.auth_token)
+            self._local.synapse = synapse
+        # Clear the Synapse cache to ensure up-to-date
+        self._local.synapse.cache.purge(after_date=0)
```

Comment:
That's intentional. I was running into some strange behaviors when the caching mechanism was enabled. I don't think it plays well with SynapseFS, which tries to manage the lifecycle of files with the `RemoteFile` class. 

That said, this would clear the cache on any computer from regular `synapseclient` usage, which is arguably an unwanted side effect. I'll try setting the cache to a new (and thus empty) temporary directory. It won't be used because I typically specify the download location of files, but at least the Synapse client won't use information from previous `synapseclient` work.

Reply:
I confirmed that caching is indeed the root cause. When caching is enabled, you can see that version 2 below is identical (3 bytes, _i.e._ `bar`) despite the local file containing `barbaz`. After deleting the `.cacheMap` file, I re-run the `synapse.store()` and you can see that version 3 now has 6 bytes as expected. 

As a compromise, I'm going to create temporary SynapseFS-specific cache directories and also purge them to avoid caching issues since I don't see another way of clearing the cache. This way, the user's usual cache (`~/.synapseCache`) is unaffected by SynapseFS. Best of both worlds. 

![image](https://user-images.githubusercontent.com/740725/212142812-986a8b37-fbe7-482e-9932-0032fc630af5.png)

Reply:
Awesome - we may want to document this particular design decision somewhere.

I read through both `IOBase` [docs](https://docs.python.org/3/library/io.html#io.IOBase) and the `RemoteFile` class, can you talk about how they are managing the lifecycle of files?

Now I'm curious on figuring out what causes this behavior in the python client (_in the future_), nothing in the `on_close` function in this PR makes it seem like the store function won't upload a new version.

Reply:
I added a link to this discussion where I purge the cache for context. 

There's something in the Synapse Python client that's causing it to upload the same unmodified file as the second version instead of picking up on the modified file and uploading that instead. 

When I say that RemoteFile is managing the lifecycle of the file, I'm referring to the fact that once the file operation is done (_e.g._ `open()` with a context manager), the file is uploaded and then deleted. This might break assumptions being made by the caching mechanism in the Synapse Python client.